gradio/utils.py

```diff
@@ -84,58 +85,83 @@ def version_check():
 
 
 def get_local_ip_address() -> str:
-    """Gets the public IP address or returns the string "No internet connection" if unable to obtain it."""
-    try:
-        ip_address = requests.get(
-            "https://checkip.amazonaws.com/", timeout=3
-        ).text.strip()
-    except (requests.ConnectionError, requests.exceptions.ReadTimeout):
-        ip_address = "No internet connection"
+    """Gets the public IP address or returns the string "No internet connection" if unable to obtain it. Does not make a new request if the IP address has already been obtained."""
+    if Context.ip_address is None:
```

Comment:
What's the performance of this approach relative to keeping `get_local_ip_address` the same and then running all subsequent telemetry requests in separate threads?

Wondering if we can avoid using the `Context` as a cache for ip address

Reply:
IIRC it shaved about 1 second for me. Is there a reason this is a bad idea? The reason I did it this way was to allow `get_local_ip_address` to run in a background thread but only have it run once per session

Reply:
Don't think it's a bad idea but if I'm reading the current code correctly, we also only call `get_ip_address` once and then just store it in the blocks state so seems like we're already caching it. So using the `Context` as a cache seems like a more complicated way to do it.

Reply:
The problem with the original approach is doesn’t easily adapt to run in a background thread. A (secondary) advantage of this née approach is that its faster when creating multiple Interfaces/Blocks in a jupyter notebook where a single Context is shared